Update docker/docker dependency to release 18.09.4

alok87 opened the pull request:
## Updates docker/docker dependency to 18.09.4

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Updates vendored files for github.com/docker/docker packages to release-branch: 18.09.4
**Justification:** https://github.com/kubernetes/kubernetes/issues/75774#issuecomment-477211119

**Which issue(s) this PR fixes**:
Fixes #75774

**Special notes for your reviewer**: 
None as of now
cc @dims 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

## Version Compatibility Discussion and Approval
https://kubernetes.slack.com/archives/C0BP8PW9G/p1554229843052600
https://github.com/kubernetes/kubernetes/pull/75843#issuecomment-481502695

## How was the SHA selected?
Active docker development happens in https://github.com/docker/docker-ce and the changes are published in the upstream docker repository https://github.com/moby/moby

Steps done to update the dependency:

- The kubernetes docker dependency was updated by finding the latest release at the time of making this PR which was - https://github.com/docker/docker-ce/releases/tag/v18.09.4
- The component which we are interested in is `engine-api` so the latest commit in 18.09.4 branch was found out by checking this https://github.com/docker/docker-ce/commits/v18.09.4/components/engine . The commit is https://github.com/docker/docker-ce/commit/5246b368bf478c808688c5b554ef0244fe9ce4f1
- Then the task was to find the same commit in the [upstream moby project](https://github.com/moby/moby) Check this https://github.com/moby/moby/commit/a588898f99d697e5ff481ecb3b273f45410f10e6
- Then the merge commit for this [commit](https://github.com/moby/moby/commit/a588898f99d697e5ff481ecb3b273f45410f10e6) was found to be https://github.com/moby/moby/commit/be7ac8be2ae072032a4005e8f232be3fc57e4127
- Kubernetes dependency was updated to https://github.com/moby/moby/commit/be7ac8be2ae072032a4005e8f232be3fc57e4127
```hack/pin-dependency.sh github.com/docker/docker be7ac8be2ae072032a4005e8f232be3fc57e4127```

alok87 commented:
/hold

alok87 commented:
/unhold

dims commented:
/ok-to-test
/hold cancel
/priority important-soon

dims commented:
/assign @dims

dims commented:
/approve
/sig node
/assign @derekwaynecarr @yujuhong @cblecker 

/hold 

(for ok from sig node folks)

cblecker commented:
/lgtm
/approve

Agree, we should have sig-node sign off on this before merge.

derekwaynecarr commented:
Does this change the version support statement for docker?  Let’s queue this up for discussion at next sig node.

alok87 commented:
/lint

alok87 commented:
/joke

yuwenma commented:
> /assign Yuwen Ma (hope is okay!)

LGTM, but I may not have permission to approve the PR.

mattjmcnaughton commented:
@alok87 thanks for this pr :)

I think you may need to update your pr now that the Go module work has been merged to replace Godep (see https://groups.google.com/forum/#!topic/kubernetes-dev/ECejidBCELI).

alok87 commented:
@mattjmcnaughton Sure, i will be pushing the rebased PR.

alok87 commented:
/retest

alok87 commented:
/assign @monopole

cblecker commented:
/unassign @monopole 

@derekwaynecarr Was this discussed in sig-node?

derekwaynecarr commented:
If I am looking at dockershim correctly, this bump shouldn’t change the min version of the api we supported per here (1.26)

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/libdocker/client.go

derekwaynecarr commented:
Ok, I can’t see a negative to this since we still should communicate with the runtime api version discovered by calling docker info, and this does not change our min version.

/approve 
/lgtm

derekwaynecarr commented:
/hold cancel

cblecker commented:
Thanks @derekwaynecarr!

/assign @liggitt 
for approval

liggitt commented on vendor/github.com/docker/docker/client/container_create.go:
```diff
@@ -43,14 +42,11 @@ func (cli *Client) ContainerCreate(ctx context.Context, config *container.Config
+	defer ensureReaderClosed(serverResp)
 	if err != nil {
-		if serverResp.statusCode == 404 && strings.Contains(err.Error(), "No such image") {
-			return response, objectNotFoundError{object: "image", id: config.Image}
```
it looks like we depend on this in https://github.com/kubernetes/kubernetes/blob/a2c200a57ab9e39be25a2599e98df807b9780818/pkg/kubelet/dockershim/libdocker/helpers.go#L165-L173

does that code need to be adjusted?

alok87 replied:
checking this.

alok87 replied:
https://github.com/moby/moby/pull/38689

alok87 replied:
changes made in https://github.com/kubernetes/kubernetes/pull/75843/commits/b1f1760486b613adfe5bfa50d1d407d12b6f3b03 please take a look

liggitt commented:
/test pull-kubernetes-godeps

liggitt commented:
one question on the NotFound behavior of create container

alok87 commented:
Can some please review the sha for docker. As i have changed it to use 18.09.4

dims commented:
/hold

for a question from @liggitt

i've applied approve label by hand (since @derekwaynecarr used `/approve` but the bot did not pick that up)

dims commented:
hmm scratch that about `/approve` this needs top level permissions. will let @liggitt do the honors!

dims commented:
@alok87 the 18.09.3->18.09.4 bump looks good

alok87 commented:
/assign @liggitt

liggitt commented:
notfound change looks good.

can you document how this sha was selected?

alok87 commented:
@liggitt I matched the last commit time at: https://github.com/docker/docker-ce/commits/v18.09.4 branch and the time at: https://github.com/moby/moby/commits/master and picked the commit from `moby/moby` based on that.

liggitt commented:
can you point to the specific commit you matched up? there are commits in https://github.com/docker/docker-ce/commits/v18.09.4 unrelated to github.com/docker/docker

alok87 commented:
Used [da823cf3a5a3437763ba12b2fa369c826d4a2bf4](https://github.com/moby/moby/commit/da823cf3a5a3437763ba12b2fa369c826d4a2bf4) based on [d14af54266dfeb55872100e28d14231b1baafe85](https://github.com/docker/docker-ce/commit/d14af54266dfeb55872100e28d14231b1baafe85)

Commit:
Docker dependency updated to 18.09.4 i.e be7ac8be2ae072032a4005e8f232be3fc57e4127

Commit:
Use docker api IsErrNotFound(). String handling is not required anymore
Why? https://github.com/kubernetes/kubernetes/pull/75843#discussion_r273785877
Moby support: https://github.com/moby/moby/pull/38689
Moby def: https://github.com/moby/moby/blob/053c6f097a1af7047e831094c0dfa3408a2c9d2b/client/errors.go#L44-L49

alok87 commented:
/retest

liggitt commented:
/lgtm
/approve

alok87 commented:
/hold cancel

liggitt commented:
/retest

alok87 commented:
/retest

yujuhong commented:
Hmm...I thinks has broken our windows test job: https://testgrid.k8s.io/google-windows#windows-gce

Kubelets couldn't come up because
```
F0415 15:17:39.377912    4916 server.go:268] failed to run Kubelet: failed to create kubelet: failed to get docker version: Error response from daemon: client version 1.40 is too new. Maximum supported API version is 1.39
```

Our Windows nodes runs Docker 18.09.3. This upates uses 18.09.4, but should still be using the 1.39 API version (https://docs.docker.com/develop/sdk/), I'm not sure why the log shows 1.40.
The linux test jobs don't seem to be affected at all.

Just as a side note, AFAIK, kubelet don't pin Docker API version even though we did consider doing that before: #53221

liggitt commented:
it appears we try to negotiate a version with the docker daemon:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/libdocker/kube_docker_client.go#L88-L96

is that failing on windows?

yujuhong commented:
> is that failing on windows?

Yes, it's failing on Windows. Although I'm not still not sure why this works on Linux...
I've sent a PR: #76621

alok87 commented:
@yujuhong i think it is picking `1.40` instead of `1.39` because of this https://github.com/alok87/kubernetes/blob/5b3a44fd56fff7caf14041231f3f52335f97312e/vendor/github.com/docker/docker/api/common.go#L6

This file was not manually changed but updating the dependency with selected SHA has changed this to `1.40` on its own.

Minimum api version was 1.39 for 18.09.4 but.
```
$ docker version
Client:
 Version:           18.09.4
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        d14af54
 Built:             Wed Mar 27 18:34:51 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.4
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       d14af54
  Built:            Wed Mar 27 18:01:48 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```
https://kubernetes.slack.com/archives/C0BP8PW9G/p1554229843052600

yujuhong commented:
Ack. The API version negotiation logic in kubelet/dockershim is wrong. I sent a fix in #76621
I didn't dig deeper to see why this works on Linux though.
